opencsp/app/sofast/lib/ProcessSofastAbstract.py

```diff
@@ -0,0 +1,296 @@
+
+    def _calculate_facet_pointing(self, reference: Literal['average'] | int = 'average') -> None:
+        """
+        Calculates facet pointing relative to the given reference.
+
+        Parameters
+        ----------
+        reference : 'average' | int
+            If 'average', the pointing reference is the average of all
+            facet pointing directions. If, int, that facet index is assumed
+            to have perfect pointing.
+        """
+        if self.data_calculation_facet is None:
+            lt.error_and_raise(ValueError, 'Slopes must be solved first by running "solve_slopes".')
+        if reference != 'average' and not isinstance(reference, int):
+            lt.error_and_raise(ValueError, 'Given reference must be int or "average".')
+        if isinstance(reference, int) and reference >= self.num_facets:
+            lt.error_and_raise(
+                ValueError, f'Given facet index, {reference:d}, is out of range of 0-{self.num_facets - 1:d}.'
+            )
+
+        # Instantiate data list
+        self.data_calculation_ensemble = []
+
+        trans_facet_ensemble_list = []
+        v_pointing_matrix = np.zeros((3, self.num_facets))
+        for idx in range(self.num_facets):
+            # Get transformation from user-input and slope solving
+            trans_1 = TransformXYZ.from_R_V(
+                self.data_ensemble_def.r_facet_ensemble[idx], self.data_ensemble_def.v_facet_locations[idx]
+            )
+            trans_2 = self.data_calculation_facet[idx].trans_alignment
+            # Calculate inverse of slope solving transform
+            trans_2 = TransformXYZ.from_V(-trans_2.V) * TransformXYZ.from_R(trans_2.R.inv())
+            # Create local to global transformation
+            trans_facet_ensemble_list.append(trans_2 * trans_1)
+
+            # Calculate pointing vector in ensemble coordinates
+            v_pointing = Vxyz((0, 0, 1)).rotate(trans_facet_ensemble_list[idx].R)
+            v_pointing_matrix[:, idx] = v_pointing.data.squeeze()
+
+        # Calculate reference pointing direction
+        if isinstance(reference, int):
+            v_pointing_ref = Vxyz(v_pointing_matrix[:, reference])
+        elif reference == 'average':
+            v_pointing_ref = Vxyz(v_pointing_matrix.mean(1))
+        # Calculate rotation to align pointing vectors
+        r_align_pointing = v_pointing_ref.align_to(Vxyz((0, 0, 1)))
+        trans_align_pointing = TransformXYZ.from_R(r_align_pointing)
+
+        # Apply alignment rotation to total transformation
+        trans_facet_ensemble_list = [trans_align_pointing * t for t in trans_facet_ensemble_list]
+
+        # Calculate global slope and surface points
+        for idx in range(self.num_facets):
+            # Get slope data
+            slopes = self.data_calculation_facet[idx].slopes_facet_xy  # facet coordinats
+
+            # Calculate surface normals in local (facet) coordinates
+            u_surf_norms = np.ones((3, slopes.shape[1]))
```

Comment:
what is at slopes.shape[1]? please consider using a enum or getter for readability

Reply:
Good point. Do you think these fixes are clear enough?